include/nanospline/BSplinePatch.h

```diff
@@ -626,13 +624,29 @@ class BSplinePatch final : public PatchBase<_Scalar, _dim>
                 p, num_samples, min_u, max_u, min_v, max_v, level);
         }
 
+        // When there are too few control points, this approach does not
+        // effectively shrink the search region.  Roll back to sampling.
+        if (num_control_points_u() <= 3 || num_control_points_v() <= 3) {
```

Comment:
I saw a commit for a bug with periodic inverse evaluation in some previous commits, do you know what the bug was there?

Reply:
The current fix is to fall back to `Base::approximate_inverse_evaluate` if the patch is periodic in any directions.  This is because `subpatch()` requires update to handle the case where the parameters are out of bound.

Reply:
Ok I see. I think for periodic patches it should be ok to use `Base::unwrap_parameter(...)` to get the appropriate parameter value and proceed normally, unless there is some algorithmic issue I am missing. I'll make an issue and fix this.

What is desired behavior for subpatch when the split point parameters are out of bounds in general? Do we want to throw an error if it is non-periodic or do we want to clamp to the nearest valid parameter `(u_min/u_max, v_min/v_max)`?

Reply:
Probably easier to explain via a picture:

![image](https://user-images.githubusercontent.com/3606672/104818624-bd093280-57f6-11eb-99ca-f8749d0c38de.png)

Imagine we have a cylinder with `u` as the periodic parameter.  The letter `A` in the image represent the region of interest, which happens to sit across periodic boundaries.  Users will specify a uv range represented by the red box to extract a `subpatch` containing `A`. Part of the red box extends beyond the patch's parametric domain, which is ok for periodic patches.  However, we cannot simply use `Base:: unwrap_parameter(...)` to map the selection box in to the domain because it will map the red box to the blue box...

Reply:
I see, thanks for clarifying. You are right that this is non-trivial; it's probably best to revert to sampling as you did.